src/detect-engine-tag.c

```diff
@@ -648,42 +647,41 @@ static int DetectTagTestPacket01 (void)
     HostInitConfig(1);
 
     SCLogDebug("running tests");
-    result = UTHGenericTest(p, 7, sigs, sid, (uint32_t *) results, 5);
+    FAIL_IF_NOT(UTHGenericTest(p, 7, sigs, sid, (uint32_t *)results, 5));
     SCLogDebug("running tests done");
 
     Host *src = HostLookupHostFromHash(&p[1]->src);
+    FAIL_IF_NULL(src);
```

Comment:
I didn't refactor the `if` block because I got a
 *tag* `re-declaration error` on line 666 because
 of line 656. The confusion now is that
 I'm not certain this line is required

Reply:
goal here is to flatten the entire function, so no branches.

Reply:
this conversion (along with the same you did for the first lines of this test) works on my end:
```diff
     Host *src = HostLookupHostFromHash(&p[1]->src);
-    if (src) {
-        void *tag = HostGetStorageById(src, host_tag_id);
-        if (tag != NULL) {
-            printf("tag should have been expired: ");
-            result = 0;
-        }
+    FAIL_IF_NULL(src);
+    FAIL_IF_NOT_NULL(HostGetStorageById(src, host_tag_id));

-        HostRelease(src);
-    }
     Host *dst = HostLookupHostFromHash(&p[1]->dst);
-    if (dst) {
-        void *tag = HostGetStorageById(dst, host_tag_id);
-        BUG_ON(tag == NULL);
+    FAIL_IF_NULL(dst);
+    void *tag = HostGetStorageById(dst, host_tag_id);
+    FAIL_IF_NULL(tag);

-        DetectTagDataEntry *iter = tag;
+    DetectTagDataEntry *iter = tag;

-        /* check internal state */
-        if (!(iter->gid == 1 && iter->sid == 2 && iter->packets == 4 && iter->count == 4)) {
-            printf("gid %u sid %u packets %u count %u: ", iter->gid, iter->sid, iter->packets, iter->cou
-            result = 0;
-        }
+    /* check internal state */
+    FAIL_IF_NOT(iter->gid == 1);
+    FAIL_IF_NOT(iter->sid == 2);
+    FAIL_IF_NOT(iter->packets == 4);
+    FAIL_IF_NOT(iter->count == 4);

-        HostRelease(dst);
-    }
-    BUG_ON(src == NULL || dst == NULL);
+    HostRelease(src);
+    HostRelease(dst);

     UTHFreePackets(p, 7);

     HostShutdown();
     TagDestroyCtx();
     StorageCleanup();
-    return result;
+    PASS;
```

Reply:
What about the `printf` statements?

Reply:
I would say that the tests failures make the most, if not all, of these print statements obsolete, since the test fails at specific points, now. But do also check Victor's comments, he understands this whole code much better ;)